### PR TITLE
Replace CQM formulation with scikit-learn plugin

### PR DESCRIPTION
**Pull request by cpotts-dwave**
First pass at replacing the CQM formulation with the scikit-learn plugin. 

I found default settings for the sliders that mirror behaviour in the original CQM with accuracy scores within about 2 decimal places. 

There's quite a few edits, all comments or suggestions welcome.

### CODE REVIEW

File: data.py [hunk 1]
```diff
@@ -68,25 +67,40 @@ def get_redundancy(self):
...
-    def solve_cqm(self, k, beta, **sampler_args):
-        """Construct and solve feature selection CQM.
+    def selected_features(self, X_new):
+        """ Post-processes result from plug-in to return features
+         Args:
+            X_new (df):
+                Reduced dataset with selected features
+
+        Returns:
+            Array of indices of selected features."""
```
Comment (hbarovertwo) [inline data.py:77]:
```suggestion
            Array of indices of selected features.
        """
```

File: README.md [hunk 2]
```diff
@@ -5,7 +5,7 @@
 # Feature Selection for CQM
 
 This demo showcases feature selection using the constrained quadratic model
-(CQM) solver.  The demo can be used with two different data sets:
+(CQM) solver via D-Wave's scikit-learn plug-in.  The demo can be used with two different data sets:
```
Comment (hbarovertwo) [inline README.md:8]: This could be a link to the github page for plugin.

File: data.py [hunk 3]
```diff
@@ -68,25 +67,40 @@ def get_redundancy(self):
         else:
             return self.calc_redundancy()
 
-    def solve_cqm(self, k, beta, **sampler_args):
-        """Construct and solve feature selection CQM.
+    def selected_features(self, X_new):
+        """ Post-processes result from plug-in to return features
+         Args:
```
Comment (hbarovertwo) [inline data.py:72]:
```suggestion
        """ Post-processes result from plug-in to return features
        
         Args:
```

File: data.py [hunk 4]
```diff
@@ -68,25 +67,40 @@ def get_redundancy(self):
         else:
             return self.calc_redundancy()
 
-    def solve_cqm(self, k, beta, **sampler_args):
-        """Construct and solve feature selection CQM.
+    def selected_features(self, X_new):
```
Comment (hbarovertwo) [inline data.py:70]: Suggest changing name to something like `get_selected_features`, seems to be a naming scheme going on here 😄

File: data.py [hunk 5]
```diff
@@ -68,25 +67,40 @@ def get_redundancy(self):
...
+        """ Post-processes result from plug-in to return features
+         Args:
+            X_new (df):
```
Comment (hbarovertwo) [inline data.py:73]: I think the type of `X_new` is `np.ndarray`

File: data.py [hunk 6]
```diff
@@ -68,25 +67,40 @@ def get_redundancy(self):
...
+        # need to iterate through and enumerate which features were selected
+        feature_names = []
+        for i in range(n):
+            for j in range(m):
+                if np.all(X_new[:, j] == self.X.iloc[:, i]):
+                    feature_names.append(i)
+                    break
+        return feature_names
+
+    def solve_feature_selection(self, k, beta):
```
Comment (hbarovertwo) [inline data.py:89]: I'm wondering if we should change `beta` to `alpha` here and in the docstring, since we end up passing this to the `alpha` argument of `SelectFromQuadraticModel`

File: data.py [hunk 7]
```diff
@@ -68,25 +67,40 @@ def get_redundancy(self):
...
-                Passed to LeapHybridCQMSampler.
+                linear and quadratic coefficients.
 
         Returns:
             Array of indices of selected features.
         """
-        alpha = beta_to_alpha(beta, k)
-        cqm = correlation_feature_selection_cqm(self.X, self.y, alpha, k)
-        return solve_feature_selection_cqm(cqm, **sampler_args)
+        X_new = SelectFromQuadraticModel(num_features=k, alpha=beta).fit_transform(self.X, self.y)
```
Comment (hbarovertwo) [inline data.py:102]:
```suggestion
        X_new = SelectFromQuadraticModel(num_features=k, alpha=beta).fit_transform(self.X.values, self.y)
```

Comment (hbarovertwo) [reply data.py:102]: I get this warning as is when running the problem:
`X has feature names, but SelectFromQuadraticModel was fitted without feature names`

File: README.md (hunk 2 above)
Comment (cpotts-dwave) [reply README.md]: Added!

File: data.py (hunk 1 above)
Comment (cpotts-dwave) [reply data.py]: Updated

File: data.py (hunk 5 above)
Comment (cpotts-dwave) [reply data.py]: Updated

File: data.py (hunk 7 above)
Comment (cpotts-dwave) [reply data.py]: Updated... warning is gone :)

File: data.py (hunk 6 above)
Comment (cpotts-dwave) [reply data.py]: Updated

File: data.py (hunk 4 above)
Comment (cpotts-dwave) [reply data.py]: Updated!

File: data.py (hunk 3 above)
Comment (cpotts-dwave) [reply data.py]: Updated

File: tests/test_data.py [hunk 8]
```diff
@@ -37,13 +35,8 @@ def test_titanic_class(self, mock_beta, mock_corr, mock_solve):
...
         self.assertTrue(np.array_equal(redundancy, from_get_redundancy))
 
-        titanic.solve_cqm(k=3, beta=0.5, time_limit=5)
-        mock_beta.assert_called_with(0.5, 3)
-        mock_corr.assert_called_with(
-            titanic.X, titanic.y, mock_beta(), 3
-        )
-        mock_solve.assert_called_with(mock_corr(), time_limit=5)
-    
+        titanic.solve_feature_selection(k=3, alpha=0.5)
```
Comment (hbarovertwo) [inline tests/test_data.py:38]: Last thing we need to address is the testing. Currently this one test file runs quite long, calling the hyrbid solver during testing is expensive. There are two things to do:

1. Mock the call to `solve_feature_selection`, following the chain on down to the hybrid solver call. This way we can test that it was called with the correct args, without actually calling it. 

2. Running the test file throws several `DeprecationWarning` stemming from dwave-system. Would be nice to address those.

Comment (cpotts-dwave) [reply tests/test_data.py:39]: 1. Updated the call to `solve_feature_selection`
2. This should be fixed in `dwave-system==1.20.0`

Also @hbarovertwo noted that the Scene tests were pretty slow at the `baseline_cv` call, so removed that portion of the test for that dataset. It remains in the Titanic tests.

**Comment by hbarovertwo** [comment]
LGTM! I don't have permissions to re request my review to hit approve.

**Comment by hbarovertwo** [comment]
May need @randomir or @arcondello to close it out

**Review by hbarovertwo — APPROVED** [review]

**Review by arcondello — APPROVED** [review]
LGTM
